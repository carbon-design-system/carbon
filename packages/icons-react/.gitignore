# TODO: remove in v11
/next
