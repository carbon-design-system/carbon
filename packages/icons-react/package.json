{
  "name": "@carbon/icons-react",
  "description": "React components for icons in digital and software products using the Carbon Design System",
  "version": "10.22.0",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "module": "es/index.js",
  "repository": {
    "type": "git",
    "url": "https://github.com/carbon-design-system/carbon.git",
    "directory": "packages/icons-react"
  },
  "bugs": "https://github.com/carbon-design-system/carbon/issues",
  "files": [
    "es",
    "lib",
    "umd"
  ],
  "keywords": [
    "ibm",
    "elements",
    "carbon",
    "carbon-elements",
    "carbon-design-system",
    "components",
    "react"
  ],
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "yarn clean && node tasks/build.js",
    "clean": "rimraf es lib umd"
  },
  "peerDependencies": {
    "react": ">=16"
  },
  "dependencies": {
    "@carbon/icon-helpers": "^10.10.0",
    "prop-types": "^15.7.2"
  },
  "devDependencies": {
    "@carbon/icon-build-helpers": "^0.12.0",
    "@carbon/icons": "^10.22.0"
  },
  "sideEffects": false
}
