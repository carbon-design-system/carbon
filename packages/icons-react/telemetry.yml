# yaml-language-server: $schema=https://unpkg.com/@ibm/telemetry-config-schema@v1/dist/config.schema.json
version: 1
projectId: e31e2f56-3767-407b-a854-ad7b9cd27677
endpoint: https://collector-prod.1am6wm210aow.us-south.codeengine.appdomain.cloud/v1/metrics
collect:
  npm:
    dependencies: null
  jsx:
    elements:
      allowedAttributeNames:
        # Icon
        - aria-hidden
        - aria-label
        - aria-labelledby
        - children
        - className
        - height
        - preserveAspectRatio
        - tabIndex
        - viewBox
        - width
        - xmlns
