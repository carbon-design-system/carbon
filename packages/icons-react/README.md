# @carbon/icons-react

> React components for icons in digital and software products using the Carbon
> Design System

## Getting started

To install `@carbon/icons-react` in your project, you will need to run the
following command using [npm](https://www.npmjs.com/):

```bash
npm install -S @carbon/icons-react
```

If you prefer [Yarn](https://yarnpkg.com/en/), use the following command
instead:

```bash
yarn add @carbon/icons-react
```

## Usage

You can import an icon component intoo your project by referring to its name:

```jsx
import { Add } from '@carbon/icons-react';
```

Each icon supports a `size` prop and supports the 16px, 20px, 24px, and 32px
sizes by default. The default size for each icon is 16px.

```jsx
import { Add } from '@carbon/icons-react';

function ExampleComponent() {
  return <Add size={32} />;
}
```

We also provide CommonJS and UMD files in the `lib` and `umd` directories,
respectively.

To import using CommonJS, you can do the following:

```js
const { Add } = require('@carbon/icons-react');
```

_Note: if you would like to find the import path for an icon, you can reference
our
[Icon Library](https://www.carbondesignsystem.com/guidelines/iconography/library)_

### Icon fill

All icons from the library support being styled by the `fill` property. You can
change the color of an icon by passing in a custom class name that sets this
property (preferred), or by passing in an inline style. For example:

```css
// CSS custom class name to set the fill of the icon to `rebeccapurple`
svg.my-custom-class {
  fill: rebeccapurple;
}
```

```jsx
import { Add } from '@carbon/icons-react';

function MyComponent() {
  return (
    <button>
      <Add aria-label="Add" className="my-custom-class" />
    </button>
  );
}
```

#### Two-tone icons

Certain icons in the library support two distinct fill colors. You can target
the inner path by using the `[data-icon-path="inner-path"]` attribute selector.
For example:

```scss
// CSS custom class name to set the fill of the icon to `yellow`
svg.my-custom-class {
  fill: yellow;
}

// Use the `data-icon-path` attribute selector to target the inner path
// where we want to set the fill to `black`. We also set `opacity` to `1` so
// that this inner-path is visible.
svg.my-custom-class [data-icon-path='inner-path'] {
  fill: black;
  opacity: 1;
}
```

```jsx
import { WarningFilled } from '@carbon/icons-react';

function MyComponent() {
  return <WarningFilled aria-label="Add" className="my-custom-class" />;
}
```

### Focus and `aria-label`

By default, the icon components from `@carbon/icons-react` are treated as
decorative content. This means that we set `aria-hidden="true"` unless certain
props are passed to the component.

If you would like the icon to be announced by a screen reader, you can supply an
`aria-label` or `aria-labelledby`. For example:

```jsx
import { Add } from '@carbon/icons-react';

function MyComponent() {
  return (
    <button>
      <Add aria-label="Add" />
    </button>
  );
}
```

Doing this will add the appropriate `role` to the `<svg>` node, as well.

If you would like the `<svg>` to receive focus, you will need to pass in a
`tabIndex` value. For example:

```jsx
import { Add } from '@carbon/icons-react';

function MyComponent() {
  return <Add aria-label="Add" tabIndex="0" />;
}
```

Including `tabIndex` and `aria-label` (or `aria-labelledby`) will set the
corresponding `tabindex` on the underlying `<svg>` and verify support in older
browsers like Internet Explorer 11 by setting `focusable` to `true`.

## üôå Contributing

We're always looking for contributors to help us fix bugs, build new features,
or help us improve the project documentation. If you're interested, definitely
check out our [Contributing Guide](/.github/CONTRIBUTING.md)! üëÄ

## üìù License

Licensed under the [Apache 2.0 License](/LICENSE).
