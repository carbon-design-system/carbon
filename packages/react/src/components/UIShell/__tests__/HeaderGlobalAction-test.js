/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { Fade } from '@carbon/icons-react';
import React from 'react';
import { mount } from 'enzyme';
import { HeaderGlobalAction } from '../';

describe('HeaderGlobalAction', () => {
  let mockProps;

  beforeEach(() => {
    mockProps = {
      'aria-label': 'Accessibility label',
      className: 'custom-class',
      onClick: jest.fn(),
      isActive: false,
    };
  });

  it('should render', () => {
    const wrapper = mount(
      <HeaderGlobalAction {...mockProps}>
        <Fade size={32} />
      </HeaderGlobalAction>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
