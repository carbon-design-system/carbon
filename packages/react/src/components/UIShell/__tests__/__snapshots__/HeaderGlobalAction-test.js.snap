// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderGlobalAction should render 1`] = `
<HeaderGlobalAction
  aria-label="Accessibility label"
  className="custom-class"
  isActive={false}
  onClick={[MockFunction]}
>
  <Button
    aria-label="Accessibility label"
    className="custom-class bx--header__action"
    disabled={false}
    hasIconOnly={true}
    iconDescription="Accessibility label"
    kind="primary"
    onClick={[MockFunction]}
    size="default"
    tabIndex={0}
    tooltipAlignment="center"
    tooltipPosition="bottom"
    type="button"
  >
    <button
      aria-label="Accessibility label"
      aria-pressed={null}
      className="custom-class bx--header__action bx--btn bx--btn--primary bx--btn--icon-only bx--tooltip__trigger bx--tooltip--a11y bx--tooltip--bottom bx--tooltip--align-center"
      disabled={false}
      onClick={[MockFunction]}
      tabIndex={0}
      type="button"
    >
      <span
        className="bx--assistive-text"
      >
        Accessibility label
      </span>
      <ForwardRef(Fade32)>
        <Icon
          fill="currentColor"
          height={32}
          preserveAspectRatio="xMidYMid meet"
          viewBox="0 0 32 32"
          width={32}
          xmlns="http://www.w3.org/2000/svg"
        >
          <svg
            aria-hidden={true}
            fill="currentColor"
            focusable="false"
            height={32}
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 32 32"
            width={32}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.24 25.14L7 26.67a13.79 13.79 0 004.18 2.44l.69-1.87A12 12 0 018.24 25.14zM4.19 18l-2 .41A14.09 14.09 0 003.86 23L5.59 22A12.44 12.44 0 014.19 18zM11.82 4.76l-.69-1.87A13.79 13.79 0 007 5.33L8.24 6.86A12 12 0 0111.82 4.76zM5.59 10L3.86 9a14.37 14.37 0 00-1.64 4.59l2 .34A12.05 12.05 0 015.59 10zM16 2V4a12 12 0 010 24v2A14 14 0 0016 2z"
            />
          </svg>
        </Icon>
      </ForwardRef(Fade32)>
    </button>
  </Button>
</HeaderGlobalAction>
`;
